### PR TITLE
Port torch/csrc/cudnn to ATen.

### PR DESCRIPTION
**Pull request by ezyang**
Just a big pile of mechanical transforms; it all mostly worked out the
way I hoped it would.

(Note that this doesn't actually PUT the cudnn source tree into ATen; that's for the next patch.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: tools/cwrap/plugins/NullableArguments.py [hunk 1]
```diff
@@ -12,3 +12,16 @@ def process_single_unpack(self, code, arg, arg_accessor):
         if 'nullable' in arg and arg['nullable']:
             return '({} == Py_None ? NULL : {})'.format(arg_accessor, code)
         return code
+
+
+class UndefinedArguments(CWrapPlugin):
```
Comment (apaszke) [inline tools/cwrap/plugins/NullableArguments.py]: Any reason why you can't replace `NullableArguments`? It's not used anywhere as far as I can see

File: torch/csrc/autograd/functions/batch_normalization.cpp [hunk 2]
```diff
@@ -73,14 +74,14 @@ auto BatchNormForward::apply(const variable_list& inputs) -> variable_list {
...
-        (THVoidTensor*)weight.unsafeGetTH(false),
-        (THVoidTensor*)bias.unsafeGetTH(false),
-        (THVoidTensor*)running_mean.unsafeGetTH(false),
-        (THVoidTensor*)running_var.unsafeGetTH(false),
-        (THVoidTensor*)save_mean.unsafeGetTH(false),
-        (THVoidTensor*)save_std.unsafeGetTH(false),
+        input_data,
+        output,
+        weight.opt_data(),
+        bias.opt_data(),
```
Comment (apaszke) [inline torch/csrc/autograd/functions/batch_normalization.cpp]: Can we please use `_var` suffixes for `at::Tensors` that are Variables? It's really hard to figure out what are the types of these things

File: torch/csrc/cudnn/BatchNorm.cpp [hunk 3]
```diff
@@ -8,40 +8,37 @@ namespace torch { namespace cudnn {
...
 }
 
-void* tensorPointer(cudnnDataType_t dataType, THVoidTensor* tensor)
+void* tensorPointer(cudnnDataType_t dataType, const at::Tensor& tensor)
 {
-  int elementSize = dataSize(dataType);
-  char* ptr = (char*) tensor->storage->data;
-  ptr += elementSize * tensor->storageOffset;
-  return ptr;
+  return tensor.data_ptr();
```
Comment (apaszke) [inline torch/csrc/cudnn/BatchNorm.cpp]: Can't we just replace all calls with `.data_ptr()`?

File: torch/csrc/cudnn/Conv.cpp [hunk 4]
```diff
@@ -482,18 +483,18 @@ Workspace chooseAlgorithm(
...
-  char* ptr = (char*) tensor->storage->data;
-  ptr += elementSize * tensor->storageOffset;
+  char* ptr = (char*) tensor.data_ptr();
 #if CUDNN_VERSION < 7000
   if (groupIdx > 0) {
     size_t size = 1;
-    for (int i = dim; i < tensor->nDimension; ++i) {
-      size *= tensor->size[i];
+    for (int i = dim; i < tensor.dim(); ++i) {
+      size *= tensor.size(i);
```
Comment (apaszke) [inline torch/csrc/cudnn/Conv.cpp]: I think that this function could be simplified. The problem before was that we had no indexing functions for `THVoidTensor`, but `at::Tensor`s should have them implemented. This function is basically a narrow along `dim`

**Review by apaszke — COMMENTED** [review]
Looks good. A few small comments

File: torch/csrc/cudnn/cuDNN.cwrap [hunk 5]
```diff
@@ -5,6 +5,8 @@
 
 #include <Python.h>
 #include "cudnn-wrapper.h"
+#include "DynamicTypes.h"
+#include "../DynamicTypes.h"
```
Comment (colesbury) [inline torch/csrc/cudnn/cuDNN.cwrap]: Are there two DynamicTypes.h files?

Comment (apaszke) [reply torch/csrc/cudnn/cuDNN.cwrap]: Yes, this PR adds a new one in `cudnn` dir

File: torch/csrc/cudnn/BatchNorm.cpp (hunk 3 above)
Comment (ezyang) [reply torch/csrc/cudnn/BatchNorm.cpp]: Yes.

File: tools/cwrap/plugins/NullableArguments.py (hunk 1 above)
Comment (ezyang) [reply tools/cwrap/plugins/NullableArguments.py]: nnwrap seems to use it. In any case, UndefinedArguments is going to get deleted soon when I have ATen take over cwrap generation.

File: torch/csrc/cudnn/cuDNN.cwrap (hunk 5 above)
Comment (ezyang) [reply torch/csrc/cudnn/cuDNN.cwrap]: Happy to rename it, I just picked the first name that came to mind. I'm hoping to delete it entirely when this code moves into ATen.

File: torch/csrc/autograd/functions/batch_normalization.cpp (hunk 2 above)
Comment (ezyang) [reply torch/csrc/autograd/functions/batch_normalization.cpp]: And to make matters worse, the implicit conversion from Variable to Tensor means you won't even notice if you get it wrong ;) (yes I'll do it)

File: torch/csrc/cudnn/Conv.cpp (hunk 4 above)
Comment (ezyang) [reply torch/csrc/cudnn/Conv.cpp]: Let me handle this when I get back from China, and don't have to go over a slow ssh connection to a GPU machine to test these changes :) (typechecker only development, aww yeahhhh)

**Comment by ezyang** [comment]
FYI, it turns out I needed to make bigger changes to cudnn to make it work with ATen (still in progress), so if other people are blocked on merging cudnn changes, they should do it (and I will spend the day remerging the changes to the new version.)
